### PR TITLE
fix: Pass response_format to LiteLLM completion for structured outputs

### PR DESCRIPTION
**Pull request by dee-walia20**
## Summary

Fixes the `LiteLLM` model so that `response_format` is actually sent to the provider when structured outputs are requested.

Previously `LiteLLM` accepted `response_format` but never added it to the request, so `output_schema` only worked via JSON-mode prompting and native structured outputs were effectively unavailable.

Changes:
- Forward `response_format` to `litellm.completion()` from `get_request_params` (covers `invoke`, `invoke_stream`, `ainvoke`, `ainvoke_stream`).
- Gate forwarding on `supports_native_structured_outputs` / `supports_json_schema_outputs` (both default `False`). They are opt-in per instance because LiteLLM fronts many providers and not all support `response_format`.
- Pass the Pydantic model / dict through to LiteLLM, which adapts it to the underlying provider, instead of hardcoding a single provider format.
- Add a structured-output cookbook example and unit tests, plus a native structured-output integration test.

## Type of change

- [x] Bug fix
- [ ] New feature
- [ ] Breaking change
- [ ] Improvement
- [x] Model update
- [ ] Other:

---

## Checklist

- [x] Code complies with style guidelines
- [x] Ran format/validation scripts (`./scripts/format.sh` and `./scripts/validate.sh`)
- [x] Self-review completed
- [x] Documentation updated (comments, docstrings)
- [x] Examples and guides: Relevant cookbook examples have been included or updated (if applicable)
- [x] Tested in clean environment
- [x] Tests added/updated (if applicable)

### Duplicate and AI-Generated PR Check

- [x] I have searched existing [open pull requests](../../pulls) and confirmed that no other PR already addresses this issue
- [ ] If a similar PR exists, I have explained below why this PR is a better approach
- [ ] Check if this PR was entirely AI-generated (by Copilot, Claude Code, Cursor, etc.)

---

## Additional Notes

Default behavior is unchanged: with both flags `False`, no `response_format` is sent and existing JSON-mode prompting is used.

### CODE REVIEW

**Comment by dee-walia20** [comment]
Hi @kausmeows , Can you look at this PR? Somehow `structured_output` doesn't work in agno while using LiteLLM.

**Comment by dee-walia20** [comment]
@ysolanky - Hey can anyone take a look this PR. We have been working with Agno + LiteLLM and this is a bug for us.

File: libs/agno/agno/models/litellm/chat.py [hunk 1]
```diff
@@ -35,6 +41,9 @@ class LiteLLM(Model):
     name: str = "LiteLLM"
     provider: str = "LiteLLM"
 
+    supports_native_structured_outputs: bool = False
+    supports_json_schema_outputs: bool = False
```
Comment (uzaxirr) [inline libs/agno/agno/models/litellm/chat.py:43]: i cant find the place where this flag is being used.

File: libs/agno/agno/models/litellm/chat.py [hunk 2]
```diff
@@ -171,21 +193,122 @@ def get_request_params(self, tools: Optional[List[Dict[str, Any]]] = None) -> Di
...
+            return {
+                "type": "json_schema",
+                "json_schema": {
+                    "name": response_format.__name__,
+                    "strict": True,
+                    "schema": schema,
+                },
+            }
+
+        return None
```
Comment (uzaxirr) [inline libs/agno/agno/models/litellm/chat.py]: See if we can somehow use the logic from ` agno/utils/models/schema_utils.py` over here.

**Review by uzaxirr — COMMENTED** [review]
Thanks for the PR

File: libs/agno/agno/models/litellm/chat.py (hunk 1 above)
Comment (dee-walia20) [reply libs/agno/agno/models/litellm/chat.py:43]: Thanks for reviewing! The supports_native_structured_outputs flag is used in all four invoke methods to conditionally pass response_format to the API:

invoke(), invoke_stream(), ainvoke(), ainvoke_stream()

The pattern in each method is:
```python
if self.supports_native_structured_outputs:
    response_format_param = self._get_response_format_param(response_format)
    if response_format_param is not None:
        completion_kwargs["response_format"] = response_format_param
```
This check is crucial because:

It prevents passing response_format to providers that don't support it (e.g., Anthropic)
LiteLLM throws UnsupportedParamsError if you pass response_format to Anthropic models
Users explicitly enable it for supported models (OpenAI, Gemini) by setting `supports_native_structured_outputs=True`

The flag is also used by the base Model class in `agno/models/base.py` to decide whether to use native structured outputs or fall back to JSON mode via prompt.

File: libs/agno/agno/models/litellm/chat.py (hunk 2 above)
Comment (dee-walia20) [reply libs/agno/agno/models/litellm/chat.py]: Yeah makes sense!
I have refactored to use `get_response_schema_for_provider` from `agno/utils/models/schema_utils.py.`

I've updated the PR with this change. Let me know if you'd like any other modifications!

**Comment by dee-walia20** [comment]
@uzaxirr- I have made changes as suggested, can you please check now?

**Comment by uzaxirr** [comment]
All tests pass! One minor issue: the file fails `ruff format --check`. Please run:

```bash
./scripts/format.sh
./scripts/validate.sh
```

After that, this is ready to merge. Approving.

**Review by uzaxirr — APPROVED** [review]

**Review by uzaxirr — APPROVED** [review]

**Comment by dee-walia20** [comment]
@uzaxirr- added format changes by ruff now, should be G2G.

**Comment by dee-walia20** [comment]
@uzaxirr- Hey, just checking if anything else is required for this PR to merge?

**Comment by nidhishgajjar** [comment]
**Orb Code Review** (powered by GLM-4.7 on [Orb Cloud](https://orbcloud.dev))

## Summary

This PR adds structured output support to LiteLLM models by introducing two new configuration flags (`supports_native_structured_outputs` and `supports_json_schema_outputs`) and implementing the `response_format` parameter conversion and passing logic. The implementation enables users to leverage LiteLLM's native structured output capabilities when supported, with comprehensive test coverage.

## Architecture

The PR extends LiteLLM functionality through a well-designed approach:

- **Safe Defaults**: Both new flags default to `False`, ensuring backward compatibility and preventing accidental breaking changes
- **Flexible Configuration**: Users can enable either or both structured output modes independently
- **Schema Conversion**: New `_get_response_format_param()` method converts Pydantic models to LiteLLM-compatible JSON schema format
- **Conditional Feature Enablement**: `response_format` parameter is only passed to LiteLLM API when `supports_native_structured_outputs=True`
- **Comprehensive Testing**: 293 lines of integration tests covering flag configuration, response format conversion, and various output schemas
- **Provider Integration**: Uses `get_response_schema_for_provider()` utility for consistent schema handling across different model providers

The implementation follows the principle of progressive enhancement - adding new capabilities without breaking existing functionality. The separation between native structured outputs and JSON schema outputs allows users to choose the appropriate mode for their use case.

## Issues

### Critical

No critical issues found. The implementation is solid and well-tested.

### Warnings

**libs/agno/agno/models/litellm/chat.py:44-45** — **flag naming could be more descriptive** — The flag names might be confusing:

```python
supports_native_structured_outputs: bool = False
supports_json_schema_outputs: bool = False
# What's the difference between "native" and "json_schema" outputs?
# Users might not understand when to use which flag
```

**Suggested fix**: Add more descriptive comments or consider renaming to clarify the distinction between the two modes.

**libs/agno/agno/models/litellm/chat.py:293-296** — **implicit dependency on schema utility** — The implementation uses `get_response_schema_for_provider` without clear documentation:

```python
if self.supports_native_structured_outputs:
    response_format_param = self._get_response_format_param(response_format)
    if response_format_param is not None:
        completion_kwargs["response_format"] = response_format_param
# Where is get_response_schema_for_provider defined?
# What are its dependencies and behavior?
```

**Suggested fix**: Document the schema utility dependency and ensure it's properly imported and available.

### Suggestions

**Missing documentation** — The PR doesn't include documentation for the new features:

```python
# No docs added for:
# - When to use supports_native_structured_outputs
# - When to use supports_json_schema_outputs
# - Examples of structured output usage
# - Differences between the two modes
```

**Suggested fix**: Add documentation to LiteLLM class docstring or create a separate guide for structured outputs.

**No examples in cookbook** — The PR adds a significant feature but doesn't include usage examples:

```python
# No cookbook examples showing:
# - How to enable structured outputs
# - Pydantic model examples
# - Response format usage
# - Comparison of different output modes
```

**Suggested fix**: Add cookbook examples demonstrating structured output usage with various response formats.

**Potential performance impact** — Structured outputs may affect token usage and latency:

```python
# Consider:
# - Does response_format increase token count?
# - What's the performance impact of schema validation?
# - Should there be warnings about performance?
# - Are there recommended use cases vs. regular outputs?
```

**Suggested fix**: Document performance characteristics and provide guidance on when to use structured outputs.

**Error handling for schema conversion** — The `_get_response_format_param` method could fail silently:

```python
def _get_response_format_param(
    self, response_format: Optional[Union[Dict, Type[BaseModel]]]
) -> Optional[Dict]:
    # ...
    if isinstance(response_format, type) and issubclass(response_format, BaseModel):
        schema = get_response_schema_for_provider(response_format, "openai")
        # What if schema conversion fails?
        # What if response_format has invalid structure?
        # No error handling visible
```

**Suggested fix**: Add try-except blocks around schema conversion with meaningful error messages.

**Model compatibility validation** — No validation that the model actually supports structured outputs:

```python
# Users can set:
model = LiteLLM(
    id="gpt-4o",
    supports_native_structured_outputs=True,  # But what if the model doesn't support this?
)
# No validation that the specified model supports the requested feature
```

**Suggested fix**: Add validation or warnings when structured outputs are requested but the model doesn't support them.

## Cross-file Impact

**Public API**: Extended with new parameters:
- `LiteLLM.supports_native_structured_outputs` - New optional parameter
- `LiteLLM.supports_json_schema_outputs` - New optional parameter
- `LiteLLM.response_format` - Enhanced parameter support

**Backward Compatibility**: Maintained through safe defaults (both flags default to False).

**Dependencies**: Uses existing `get_response_schema_for_provider()` utility.

**Test Coverage**: Excellent - 293 lines of comprehensive integration tests added.

**Breaking Changes**: None. This is a pure enhancement.

**Documentation**: Limited to code comments and test docstrings.

## Assessment

✅ Approve

This PR is a well-implemented feature addition that extends LiteLLM capabilities in a safe, backward-compatible manner. The use of safe defaults (both flags default to False) ensures existing code continues to work without modification. The comprehensive test coverage demonstrates thorough validation of the new functionality.

The architecture is clean with good separation of concerns:
- Clear configuration flags for different structured output modes
- Robust schema conversion logic
- Conditional feature enablement that only affects behavior when explicitly requested
- Comprehensive test coverage that validates both success and edge cases

While minor improvements could be made (better documentation, cookbook examples, performance guidance), these are not blocking concerns. The core implementation is solid, well-tested, and follows existing code patterns in the codebase.

This feature enables valuable functionality for users who need structured outputs from LiteLLM models, and the implementation provides the flexibility and safety needed for production use.

**Review by harshsinha03 — APPROVED** [review]
LGTM! Thanks for contributing
